### PR TITLE
feat: Require SCA for registration

### PR DESCRIPTION
**Pull request by m-horky**
* Card ID: CCT-612

### CODE REVIEW

File: test/rhsmlib/services/test_register.py [hunk 1]
```diff
@@ -315,33 +293,11 @@ def test_register_normally_no_owner(self, mock_persist_consumer, mock_write_cach
...
-            mock.call("pre_register_consumer", name="name", facts={}),
-            mock.call("post_register_consumer", consumer=expected_consumer, facts={}),
-        ]
-        self.assertEqual(expected_plugin_calls, self.mock_pm.run.call_args_list)
-        assert "owner" not in consumer
+        with pytest.raises(
+            exceptions.ServiceError,
+            match="Registration is only possible when the organization is in Simple Content Access Mode.",
+        ):
+            register_service.register("org", name="name", environments=["environment"])
```
[the remark this answers is not in the document]
Comment (m-horky) [reply test/rhsmlib/services/test_register.py:300]: I don't know. I'll update it.

[the remark this answers is not in the document]
Comment (m-horky) [reply test/rhsmlib/services/test_register.py:300]: I believe you are always part of some organization (though I'm not sure about cloud registration v2).

**Comment by m-horky** [comment]
@jirihnidek The no-org state is already handled on higher levels. For both CLI and D-Bus, the `determine_owner_key` has a `no_owner_cb` callback that stops the registration.

CLI:

https://github.com/candlepin/subscription-manager/blob/a01dabd2b6f8f6078661b671c5431d601c8171ea/src/subscription_manager/cli_command/register.py#L347-L349
https://github.com/candlepin/subscription-manager/blob/a01dabd2b6f8f6078661b671c5431d601c8171ea/src/subscription_manager/cli_command/register.py#L496-L501

D-Bus:

https://github.com/candlepin/subscription-manager/blob/a01dabd2b6f8f6078661b671c5431d601c8171ea/src/rhsmlib/dbus/objects/register.py#L203-L207
https://github.com/candlepin/subscription-manager/blob/a01dabd2b6f8f6078661b671c5431d601c8171ea/src/rhsmlib/dbus/objects/register.py#L253-L258

**Comment by jirihnidek** [comment]
It looks OK to me, but it would be to modify cockpit integration tests ... I can see that you already try to do it here: https://github.com/candlepin/subscription-manager-cockpit/pull/70 but some tests are still failing...

**Comment by m-horky** [comment]
I've tried to fix the Cockpit tests, but it got to the limits of the timebox I set up for it, see downstream CCT-624 for details.

File: src/rhsmlib/services/register.py [hunk 2]
```diff
@@ -132,6 +132,16 @@ def register(
...
             cp_provider.close_all_connections()
 
+        access_mode: str = consumer.get("owner", {}).get("contentAccessMode", "unknown")
+        if access_mode != "org_environment":
+            log.error(
+                f"Organization's content access mode is '{access_mode}'. "
+                "Only simple content access/'org_environment' is allowed."
+            )
+            raise exceptions.ServiceError(
+                _("Registration is only possible when the organization is in Simple Content Access Mode.")
```
Comment (ptoscano) [inline src/rhsmlib/services/register.py:142]: a minor wording note, according to what I see in other user-facing messages: s/Simple Content Access Mode/Simple Content Access (SCA) mode/
- lowercase "mode" (it is not part of the acronym
- add "(SCA)" as acronym, so it can be spotted even by those that don't know the "expanded name"

File: src/rhsmlib/services/register.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def register(
         service_level: str = None,
         usage: str = None,
         jwt_token: str = None,
-        **kwargs: dict
+        **kwargs: dict,
```
Comment (ptoscano) [inline src/rhsmlib/services/register.py:52]: hm extra/unrelated change?

File: test/rhsmlib/services/test_register.py [hunk 4]
```diff
@@ -271,35 +273,11 @@ def test_register_normally_old_candlepin(self, mock_persist_consumer, mock_write
...
-        mock_write_cache.assert_called_once()
-        expected_plugin_calls = [
-            mock.call("pre_register_consumer", name="name", facts={}),
-            mock.call("post_register_consumer", consumer=expected_consumer, facts={}),
-        ]
-        self.assertEqual(expected_plugin_calls, self.mock_pm.run.call_args_list)
-        assert "owner" in consumer
-        assert "contentAccessMode" in consumer["owner"]
-        assert "entitlement" == consumer["owner"]["contentAccessMode"]
+        with pytest.raises(
```
Comment (ptoscano) [inline test/rhsmlib/services/test_register.py:276]: since the tests use unittest and not pytest, I'd avoid adding also pytest here; please use `self.assertRaises()` instead

File: test/rhsmlib/services/test_register.py [hunk 5]
```diff
@@ -315,33 +293,11 @@ def test_register_normally_no_owner(self, mock_persist_consumer, mock_write_cach
...
-
-        mock_persist_consumer.assert_called_once_with(expected_consumer)
-        mock_write_cache.assert_called_once()
-        expected_plugin_calls = [
-            mock.call("pre_register_consumer", name="name", facts={}),
-            mock.call("post_register_consumer", consumer=expected_consumer, facts={}),
-        ]
-        self.assertEqual(expected_plugin_calls, self.mock_pm.run.call_args_list)
-        assert "owner" not in consumer
+        with pytest.raises(
```
Comment (ptoscano) [inline test/rhsmlib/services/test_register.py:296]: ditto (pytest -> unittest)

**Review by ptoscano — CHANGES_REQUESTED** [review]
generally good, few things to change

File: src/rhsmlib/services/register.py [hunk 6]
```diff
@@ -132,6 +132,19 @@ def register(
...
+        if access_mode != "org_environment":
+            log.error(
+                f"Organization's content access mode is '{access_mode}'. "
+                "Only simple content access/'org_environment' is allowed."
+            )
+            raise exceptions.ServiceError(
+                _(
+                    "Registration is only possible when the organization "
+                    "is in Simple Content Access (SCA) mode."
+                )
```
Comment (ptoscano) [inline src/rhsmlib/services/register.py:145]: (I thought I had written that, must had lost it...) One important thing to do here is to actually unregister this consumer right before raising the failure, otherwise there will be a "stale" consumer in Candlepin that can consume subscriptions.

**Comment by ptoscano** [comment]
Thank you for all the updates, now it LGTM. After all the cockpit tests were fixed/adapted, now this works for cockpit too :D

@jirihnidek would you like to take a second look at it, after the latest changes?

**Review by jirihnidek — APPROVED** [review]
LGTM
